src/Formatter/IntlFormatter.php

```diff
@@ -0,0 +1,185 @@
+
+declare(strict_types=1);
+
+namespace Brick\DateTime\Formatter;
+
+use Brick\DateTime\LocalDate;
+use Brick\DateTime\LocalDateTime;
+use Brick\DateTime\LocalTime;
+use Brick\DateTime\ZonedDateTime;
+use DateTimeZone;
+use IntlDateFormatter;
+use IntlDatePatternGenerator;
+
+use function array_keys;
+use function extension_loaded;
+use function get_class;
+use function in_array;
+use function sprintf;
+use function str_split;
+
+use const PHP_VERSION_ID;
+
+/**
+ * Formats the value using the Intl extension.
+ */
+final class IntlFormatter implements DateTimeFormatter
+{
+    public const FULL = IntlDateFormatter::FULL;
+    public const LONG = IntlDateFormatter::LONG;
+    public const MEDIUM = IntlDateFormatter::MEDIUM;
+    public const SHORT = IntlDateFormatter::SHORT;
+
+    private string $locale;
+
+    private int $dateFormat;
+
+    private int $timeFormat;
+
+    private string $pattern;
+
+    private function __construct(string $locale, int $dateFormat, int $timeFormat, string $pattern)
+    {
+        $this->locale = $locale;
+        $this->dateFormat = $dateFormat;
+        $this->timeFormat = $timeFormat;
+        $this->pattern = $pattern;
+
+        if (! extension_loaded('intl')) {
+            throw new DateTimeFormatException('IntlFormatter requires ext-intl to be installed and enabled.');
+        }
+    }
+
+    /**
+     * Returns a formatter of given type for a date value.
+     */
+    public static function ofDate(string $locale, int $format): self
+    {
+        return new self($locale, $format, IntlDateFormatter::NONE, '');
+    }
+
+    /**
+     * Returns a formatter of given type for a time value.
+     */
+    public static function ofTime(string $locale, int $format): self
+    {
+        return new self($locale, IntlDateFormatter::NONE, $format, '');
+    }
+
+    /**
+     * Returns a formatter of given type for a date-time value.
+     */
+    public static function ofDateTime(string $locale, int $dateFormat, int $timeFormat): self
+    {
+        return new self($locale, $dateFormat, $timeFormat, '');
+    }
+
+    /**
+     * Returns a formatter with given ICU SimpleFormat pattern.
+     */
+    public static function ofPattern(string $locale, string $pattern): self
+    {
+        return new self($locale, IntlDateFormatter::NONE, IntlDateFormatter::NONE, $pattern);
+    }
+
+    /**
+     * Returns a formatter with a pattern that best matches given skeleton.
+     */
+    public static function ofSkeleton(string $locale, string $skeleton): self
+    {
+        if (PHP_VERSION_ID < 80100) {
+            throw new DateTimeFormatException('IntlFormatter::ofSkeleton() is only available in PHP 8.1 and above.');
+        }
+
+        $generator = new IntlDatePatternGenerator($locale);
+        $pattern = $generator->getBestPattern($skeleton);
+
+        if ($pattern === false) {
+            throw new DateTimeFormatException('Failed to resolve the best formatting pattern for given locale and skeleton.');
```

Comment:
Passing locale and pattern in message that throw this exception could be useful I guess